### PR TITLE
ch4/rma: Implement accumulate ordering on OFI

### PR DESCRIPTION
**Pull request by shawnccx**
Accumulate ordering on OFI level is implemented as follows:

- Acc, Get_acc, Fetch_and_op and Compare_and_swap are supported.
- Large message is chunked into smaller messages for which OFI provider
guarantees ordering.
- If max ordering size form OFI provider is smaller than basic data
type, then ordering is not guaranteed.
- In order to handle outstanding active message acc ops: an OFI acc op
cannot be issued until all active message acc ops are done, and vice versa.

### CODE REVIEW

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Comment by shawnccx** [comment]
Since all tests passed, any comments on this PR?

**Comment by raffenet** [comment]
@minsii can you review this?

**Comment by minsii** [comment]
@raffenet : Yes. It is in my todo list. I will review it later today.

File: src/mpid/ch4/netmod/ofi/ofi_rma.h [hunk 1]
```diff
@@ -1043,7 +1061,19 @@ static inline int MPIDI_OFI_do_get_accumulate(const void *origin_addr,
     if (max_count == 0)
         goto am_fallback;
 
-    max_size = MPL_MIN(max_count * dt_size, MPIDI_Global.max_write / dt_size * dt_size);
+    max_size = MPIDI_OFI_check_acc_order_size(win, max_count * dt_size);
+    max_size = MPL_MIN(max_size, MPIDI_Global.max_write);
+    /* round down to multiple of dt_size */
+    max_size = max_size / dt_size * dt_size;
+    /* It's impossible to chunk data if buffer size is smaller than basic datatype size. */
+    MPIR_Assert(max_size >= dt_size);
```
Comment (minsii) [inline src/mpid/ch4/netmod/ofi/ofi_rma.h]: Should here set goto am_fallback if false ?

File: test/mpi/rma/acc_ordering.c [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    MPI_Win_free(&win);
+
+    /* 2. None ordering */
+    MPI_Info_create(&info_in);
+    MPI_Info_set(info_in, "accumulate_ordering", "none");
+    MPI_Win_create(data, sizeof(twoint_t) * ARRAY_LEN, 1, info_in, MPI_COMM_WORLD, &win);
+    MPI_Info_free(&info_in);
+
+    if (me == 0) {
+        memset((void*)data, 0, ARRAY_LEN * sizeof(twoint_t));
```
Comment (minsii) [inline test/mpi/rma/acc_ordering.c]: You need to add a `fence` before this `memset`. Any access to the window memory has to be wrapped by a epoch. Or you can reset it before `win_create`.

File: test/mpi/rma/acc_ordering.c [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        verify_nondeterministic_result(data, ARRAY_LEN, expected, "Large array test case for none ordering");
+    }
+
+    MPI_Win_free(&win);
+
+    /* 3. Mix MAX_LOC/MIN_LOC and MPI_REPLACE */
+    MPI_Win_create(data, sizeof(twoint_t) * ARRAY_LEN, 1, MPI_INFO_NULL, MPI_COMM_WORLD, &win);
+
+    if (me == 0) {
+        memset((void*)data, 0, ARRAY_LEN * sizeof(twoint_t));
```
Comment (minsii) [inline test/mpi/rma/acc_ordering.c]: Same here.

File: src/mpid/ch4/netmod/ofi/ofi_init.h [hunk 4]
```diff
@@ -1618,7 +1621,8 @@ static inline int MPIDI_OFI_init_hints(struct fi_info *hints)
      * else (AM mode) delivery complete is not required */
     if(MPIDI_OFI_ENABLE_RMA)
         hints->tx_attr->op_flags |= FI_DELIVERY_COMPLETE;
-    hints->tx_attr->msg_order = FI_ORDER_SAS;
+    /* Apply most restricted msg order in hints for RMA.*/
+    hints->tx_attr->msg_order = FI_ORDER_SAS | FI_ORDER_RAR | FI_ORDER_RAW | FI_ORDER_WAR | FI_ORDER_WAW;
```
[the remark this answers is not in the document]
Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_init.h]: Send and RMA read/write are not working together in any case. Send is used in active message level and read/write is used in OFI level. We try to give OFI minimum hints (or, limits) to let it enhance performance if possible.

File: src/mpid/ch4/netmod/ofi/ofi_rma.h [hunk 5]
```diff
@@ -749,8 +749,14 @@ static inline int MPIDI_NM_mpi_compare_and_swap(const void *origin_addr,
     rbuffer = (char *) result_addr + true_lb;
     tbuffer = (void *) (MPIDI_OFI_winfo_base(win, target_rank) + offset);
 
-    max_size = MPIDI_OFI_QUERY_COMPARE_ATOMIC_COUNT;
-    MPIDI_OFI_query_datatype(datatype, &fi_dt, MPI_OP_NULL, &fi_op, &max_size, &dt_size);
+    max_count = MPIDI_OFI_QUERY_COMPARE_ATOMIC_COUNT;
+    MPIDI_OFI_query_datatype(datatype, &fi_dt, MPI_OP_NULL, &fi_op, &max_count, &dt_size);
+    max_size = MPIDI_OFI_check_acc_order_size(win, dt_size);
+    /* It's impossible to transfer data if buffer size is smaller than basic datatype size. */
+    MPIR_Assert(max_size >= dt_size);
```
[the remark this answers is not in the document]
Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_rma.h]: Our original thought is to let user there is no way for OFI to guarantee order in such basic data size. I'll change to fall back to AM and for the below two.

File: test/mpi/rma/acc_ordering.c (hunk 2 above)
Comment (shawnccx) [reply test/mpi/rma/acc_ordering.c]: I will reset it before `win_create`.

File: test/mpi/rma/acc_ordering.c (hunk 3 above)
Comment (shawnccx) [reply test/mpi/rma/acc_ordering.c]: I will reset it before `win_create`.

**Comment by shawnccx** [comment]
Comments have been addressed in latest commits.

**Comment by minsii** [comment]
test:jenkins/ch4/most

**Comment by minsii** [comment]
test:jenkins/ch4/most

File: src/mpid/ch4/netmod/ofi/ofi_rma.h [hunk 6]
```diff
@@ -771,6 +778,14 @@ static inline int MPIDI_NM_mpi_compare_and_swap(const void *origin_addr,
     return mpi_errno;
   fn_fail:
     goto fn_exit;
+  am_fallback:
+    if (MPIDI_CH4U_WIN(win, info_args).accumulate_ordering & (MPIDI_CH4I_ACCU_ORDER_WAW | MPIDI_CH4I_ACCU_ORDER_WAR)) {
+        /* Wait for OFI acc to complete.
+         * For now, there is no FI flag to track atomic only ops, we use RMA level cntr. */
```
Comment (minsii) [inline src/mpid/ch4/netmod/ofi/ofi_rma.h]: Should this also include `MPIDI_CH4I_ACCU_ORDER_RAW` ? And please also replace `acc` with `cas` in the comment.

Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_rma.h]: We only have counters for all `acc` ops, here `cas` belongs `acc`.

**Comment by minsii** [comment]
@shawnccx I have only one new comment. Everything else looks fine.
@raffenet It looks like I cannot trigger the jenkins tests again.. Could you please trigger it for me ?

File: src/mpid/ch4/netmod/ofi/ofi_rma.h (hunk 6 above)
Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_rma.h]: I will make comment more clear there, and add `MPIDI_CH4I_ACCU_ORDER_RAW`. Also, I will add `void` for `max_size` in `comp_and_swap` to eliminate `unused variable` warning in default mode.

Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_rma.h]: Oh, my bad, it is `ofi` level fence, so the comment below has explained that there is no cas only cntr.

**Comment by shawnccx** [comment]
@minsii Comment addressed in new commits.

**Comment by minsii** [comment]
test:jenkins/ch4/most

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Pull request merged**